# Data Loading

In [2]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [3]:
import numpy as np
import pandas as pd

In [26]:
df = pd.read_csv("/content/drive/MyDrive/price_Onion.csv")

# Data Preprocessing

In [27]:
df.head()


State State_Code  District  ...  Max_Price Modal_Price         Date
0  Karnataka         KK  Bagalkot  ...       2200      1700.0  30 Oct 2021
1  Karnataka         KK  Bagalkot  ...       2000      1500.0  16 Oct 2021
2  Karnataka         KK  Bagalkot  ...       2000      1600.0  09 Oct 2021
3  Karnataka         KK  Bagalkot  ...       2350      1800.0  28 Sep 2021
4  Karnataka         KK  Bagalkot  ...       2000      1700.0  21 Sep 2021

[5 rows x 14 columns]

In [28]:
df.shape

(61083, 14)

**Removing Duplicate entries**

In [29]:
df = df.drop_duplicates()

In [30]:
df.shape

(59281, 14)

Initially there was 61083 entries, now after removing duplicated entries there are 59281 entries.

In [54]:
df.columns

Index(['State', 'State_Code', 'District', 'District_Code', 'Market',
       'Market_Code', 'Commodity', 'Commodity_Code', 'Variety', 'Grade',
       'Min_Price', 'Max_Price', 'Modal_Price', 'Date'],
      dtype='object')

**Droping some columns which are not useful going ahead**

In [55]:
df = df.drop(['State', 'State_Code',
        'Commodity', 'Commodity_Code', 'Variety', 'Grade'], axis=1)

In [56]:
df.head()

District  District_Code     Market  ...  Max_Price  Modal_Price         Date
0  Bagalkot             25  Bagalakot  ...       2200       1700.0  30 Oct 2021
1  Bagalkot             25  Bagalakot  ...       2000       1500.0  16 Oct 2021
2  Bagalkot             25  Bagalakot  ...       2000       1600.0  09 Oct 2021
3  Bagalkot             25  Bagalakot  ...       2350       1800.0  28 Sep 2021
4  Bagalkot             25  Bagalakot  ...       2000       1700.0  21 Sep 2021

[5 rows x 8 columns]

In [57]:
df.describe()

District_Code   Market_Code     Min_Price     Max_Price   Modal_Price
count   59281.000000  59281.000000  59281.000000  59281.000000  59281.000000
mean        9.761880    388.339350    888.795095   1600.596026   1258.869967
std         8.205651    493.426972    802.753883   1237.757555    948.130330
min         1.000000    107.000000      0.000000      0.000000      0.000000
25%         2.000000    109.000000    400.000000    900.000000    700.000000
50%         7.000000    120.000000    665.000000   1200.000000   1000.000000
75%        15.000000    735.000000   1000.000000   1983.000000   1500.000000
max        28.000000   4400.000000  14000.000000  42000.000000  17000.000000

In [58]:
df['Modal_Price'].describe()

count    59281.000000
mean      1258.869967
std        948.130330
min          0.000000
25%        700.000000
50%       1000.000000
75%       1500.000000
max      17000.000000
Name: Modal_Price, dtype: float64

 **Districts**

In [59]:
df['District'].describe()

count         59281
unique           25
top       Bangalore
freq          11645
Name: District, dtype: object

There are 25 unique districts.
While the Bangalore district has the maximum number of entries.

In [60]:
district_names =df.District.unique()

In [61]:
district_names

array(['Bagalkot', 'Bangalore', 'Belgaum', 'Bellary', 'Bidar', 'Bijapur',
       'Chamrajnagar', 'Chikmagalur', 'Chitradurga', 'Davangere',
       'Dharwad', 'Gadag', 'Gulbarga', 'Hassan', 'Haveri',
       'Karwar(Uttar Kannad)', 'Kolar', 'Madikeri(Kodagu)', 'Mandya',
       'Mangalore(Dakshin Kannad)', 'Mysore', 'Raichur', 'Shimoga',
       'Tumkur', 'Udupi'], dtype=object)

In [62]:
df['Market'].describe()

count         59281
unique           94
top       Bangalore
freq           9576
Name: Market, dtype: object

In [63]:
market_names = df.Market.unique()

In [64]:
market_names

array(['Bagalakot', 'Jamakhandi', 'Mhalingapur', 'Bangalore',
       'Channapatana', 'Doddaballa Pur', 'Hoskote', 'Kanakapura',
       'Ramanagara', 'Bailahongal', 'Belgaum', 'Nippani', 'Ramdurga',
       'Sankeshwar', 'Bellary', 'H.B. Halli', 'Hospet', 'Bhalki', 'Bidar',
       'Humanabad', 'Bijapur', 'Sindagi', 'Chamaraj Nagar', 'Gundlupet',
       'Kollegal', 'Bagepalli', 'Chikkamagalore', 'Kadur', 'Koppa',
       'Moodigere', 'Sringeri', 'Tarikere', 'Challakere', 'Holalkere',
       'Davangere', 'Harappana Halli', 'Harihara', 'Honnali', 'Jagalur',
       'Annigeri', 'Dharwar', 'Hubli (Amaragol)', 'Gadag', 'Gulbarga',
       'Arakalgud', 'Arasikere', 'Belur', 'Channarayapatna', 'Hassan',
       'Holenarsipura', 'Sakaleshpura', 'Haveri', 'Ranebennur', 'Karwar',
       'Siddapur', 'Sirsi', 'Bangarpet', 'Chickkaballapura', 'Chintamani',
       'Gowribidanoor', 'Kolar', 'Malur', 'Mulabagilu', 'Srinivasapur',
       'Gonikappal', 'Madikeri', 'Somvarpet', 'K.R. Pet', 'Maddur',
       'Man

**Number of entries for every district**

In [65]:
number_of_entries ={}
for district in district_names:
  number_of_entries[district] =0

In [76]:
df = df.reset_index(drop=True)

In [77]:
df.shape

(59281, 8)

In [82]:

for i in range(df.shape[0]):
  for district in district_names:
    if df['District'][i] == district:
      number_of_entries[district] +=1


In [83]:
number_of_entries

{'Bagalkot': 60393,
 'Bangalore': 70926,
 'Belgaum': 65028,
 'Bellary': 61202,
 'Bidar': 59448,
 'Bijapur': 60280,
 'Chamrajnagar': 61071,
 'Chikmagalur': 62743,
 'Chitradurga': 59613,
 'Davangere': 62931,
 'Dharwad': 67075,
 'Gadag': 60151,
 'Gulbarga': 59291,
 'Hassan': 61695,
 'Haveri': 60043,
 'Karwar(Uttar Kannad)': 59555,
 'Kolar': 64157,
 'Madikeri(Kodagu)': 59309,
 'Mandya': 59600,
 'Mangalore(Dakshin Kannad)': 59604,
 'Mysore': 62490,
 'Raichur': 62284,
 'Shimoga': 61011,
 'Tumkur': 60938,
 'Udupi': 60468}

In [84]:
number_of_entries_market_wise ={}
for market in market_names:
  number_of_entries_market_wise[market] =0

In [85]:
for i in range(df.shape[0]):
  for market in market_names:
    if df['Market'][i] == market:
      number_of_entries_market_wise[market] +=1

In [86]:
number_of_entries_market_wise

{'Annigeri': 3,
 'Arakalgud': 8,
 'Arasikere': 727,
 'Bagalakot': 645,
 'Bagepalli': 408,
 'Bailahongal': 13,
 'Bangalore': 9576,
 'Bangarpet': 168,
 'Belgaum': 5674,
 'Bellary': 62,
 'Belur': 31,
 'Bhadravathi': 187,
 'Bhalki': 3,
 'Bidar': 71,
 'Bijapur': 996,
 'Challakere': 100,
 'Chamaraj Nagar': 601,
 'Channapatana': 53,
 'Channarayapatna': 175,
 'Chickkaballapura': 1313,
 'Chikkamagalore': 1944,
 'Chintamani': 389,
 'Davangere': 3263,
 'Dharwar': 109,
 'Doddaballa Pur': 920,
 'Gadag': 870,
 'Gonikappal': 8,
 'Gowribidanoor': 538,
 'Gubbi': 2,
 'Gulbarga': 10,
 'Gundlupet': 1139,
 'H.B. Halli': 3,
 'Harappana Halli': 2,
 'Harihara': 6,
 'Hassan': 1451,
 'Haveri': 71,
 'Holalkere': 232,
 'Holenarsipura': 15,
 'Honnali': 370,
 'Hoskote': 41,
 'Hospet': 1856,
 'Hubli (Amaragol)': 7682,
 'Humanabad': 93,
 'Hunsur': 27,
 'Jagalur': 9,
 'Jamakhandi': 450,
 'K.R. Pet': 220,
 'Kadur': 557,
 'Kanakapura': 317,
 'Karwar': 2,
 'Kolar': 2028,
 'Kollegal': 50,
 'Koppa': 17,
 'Kundapura': 334,


**Exploring bangalore district**

In [87]:
df_bangalore = df.loc[df['District'] == "Bangalore"]

In [94]:
df_bangalore.shape

(11645, 8)

In [104]:
df_bangalore = df_bangalore.reset_index(drop=True)

In [93]:
df_bangalore.head()

District  District_Code     Market  ...  Max_Price  Modal_Price         Date
0  Bangalore              1  Bangalore  ...       2500       2000.0  30 Oct 2021
1  Bangalore              1  Bangalore  ...       2200       2100.0  30 Oct 2021
2  Bangalore              1  Bangalore  ...       2500       2000.0  29 Oct 2021
3  Bangalore              1  Bangalore  ...       2200       2100.0  29 Oct 2021
4  Bangalore              1  Bangalore  ...       2500       2000.0  28 Oct 2021

[5 rows x 8 columns]

In [96]:
df_bangalore['Date'].describe()

count           11645
unique           3681
top       28 Jan 2015
freq               11
Name: Date, dtype: object

In [111]:
df_bangalore.head()

District  District_Code     Market  ...  Max_Price  Modal_Price         Date
0  Bangalore              1  Bangalore  ...       2500       2000.0  30 Oct 2021
1  Bangalore              1  Bangalore  ...       2200       2100.0  30 Oct 2021
2  Bangalore              1  Bangalore  ...       2500       2000.0  29 Oct 2021
3  Bangalore              1  Bangalore  ...       2200       2100.0  29 Oct 2021
4  Bangalore              1  Bangalore  ...       2500       2000.0  28 Oct 2021

[5 rows x 8 columns]

In [108]:
df_bangalore['Date'][0][3:6]

'Oct'

In [107]:
dates=[]
months=[]

for i in range(df_bangalore.shape[0]):
  dates.append(int(df_bangalore['Date'][i][:2]))
  months.append(df_bangalore['Date'][i][3:6])

In [112]:
months[:6]

['Oct', 'Oct', 'Oct', 'Oct', 'Oct', 'Oct']

In [113]:
dates[:6]

[30, 30, 29, 29, 28, 28]

In [114]:
for i in range(len(months)):
  if months[i] == "Jan":
    months[i] = 1
  elif months[i] == "Feb":
    months[i] = 2
  elif months[i] == "Mar":
    months[i] = 3
  elif months[i] == "Apr":
    months[i] = 4
  elif months[i] == "May":
    months[i] = 5
  elif months[i] == "Jun":
    months[i] = 6
  elif months[i] == "Jul":
    months[i] = 7
  elif months[i] == "Aug":
    months[i] = 8
  elif months[i] == "Sep":
    months[i] = 9
  elif months[i] == "Oct":
    months[i] = 10
  elif months[i] == "Nov":
    months[i] = 11
  elif months[i] == "Dec":
    months[i] = 12

    



In [117]:
df_bangalore['Day'] = dates
df_bangalore["Month"] = months

In [119]:
df_bangalore.head()

District  District_Code     Market  ...         Date  Day  Month
0  Bangalore              1  Bangalore  ...  30 Oct 2021   30     10
1  Bangalore              1  Bangalore  ...  30 Oct 2021   30     10
2  Bangalore              1  Bangalore  ...  29 Oct 2021   29     10
3  Bangalore              1  Bangalore  ...  29 Oct 2021   29     10
4  Bangalore              1  Bangalore  ...  28 Oct 2021   28     10

[5 rows x 10 columns]

In [121]:
price=[]
for i in range(df_bangalore.shape[0]):
  price.append(df_bangalore['Modal_Price'][i]//100)

In [123]:
df_bangalore['Price'] = price 

In [125]:
df_bangalore.head()

District  District_Code     Market  ...  Day  Month  Price
0  Bangalore              1  Bangalore  ...   30     10   20.0
1  Bangalore              1  Bangalore  ...   30     10   21.0
2  Bangalore              1  Bangalore  ...   29     10   20.0
3  Bangalore              1  Bangalore  ...   29     10   21.0
4  Bangalore              1  Bangalore  ...   28     10   20.0

[5 rows x 11 columns]

In [129]:
X = df_bangalore.values[:,[3,8,9]]
Y = df_bangalore.values[:,10]

In [130]:
X

array([[107, 30, 10],
       [107, 30, 10],
       [107, 29, 10],
       ...,
       [773, 7, 1],
       [773, 5, 1],
       [773, 4, 1]], dtype=object)

In [131]:
 Y

array([20.0, 21.0, 20.0, ..., 10.0, 11.0, 11.0], dtype=object)

**Train Test Split**

In [132]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split( X, Y, test_size=0.2, random_state=0)

# Model Fitting

**Linear Regression**

In [154]:
from sklearn.linear_model import LinearRegression
rgsr = LinearRegression()
rgsr.fit(X_train,Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [155]:
y_pred= rgsr.predict(X_test)

In [156]:
y_pred[:10]

array([13.28966646, 14.08207048, 15.7045969 , 16.55986486, 17.79943521,
       13.28966646, 10.72386256, 16.8813033 , 15.62287378, 12.37782095])

In [157]:
Y_test[:10]

array([25.0, 10.0, 9.0, 10.0, 22.0, 18.0, 3.0, 20.0, 28.0, 18.0],
      dtype=object)

In [158]:
from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(Y_test, y_pred))  
print('MSE:', metrics.mean_squared_error(Y_test, y_pred))  
print('RMSE:', np.sqrt(metrics.mean_squared_error(Y_test, y_pred)))

MAE: 6.826446674471035
MSE: 101.58740405464822
RMSE: 10.079057696761549


**Lasso Regression**

In [159]:

from sklearn import linear_model
rgsrL = linear_model.Lasso(alpha=0.1)
rgsrL.fit(X_train,Y_train)

Lasso(alpha=0.1, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

In [160]:
y_pred= rgsrL.predict(X_test)


In [162]:
y_pred[:10]

array([13.21317147, 14.00129766, 15.57755004, 16.36567623, 17.52656239,
       13.21317147, 10.8487929 , 16.7384362 , 15.57755004, 12.42504528])

In [161]:
print('MAE:', metrics.mean_absolute_error(Y_test, y_pred))  
print('MSE:', metrics.mean_squared_error(Y_test, y_pred))  
print('RMSE:', np.sqrt(metrics.mean_squared_error(Y_test, y_pred)))

MAE: 6.827195410048399
MSE: 101.65512658727093
RMSE: 10.082416703711017


**SVM**

In [170]:
from sklearn import svm
rgsrSVM = svm.SVR()
rgsrSVM.fit(X_train,Y_train)
y_pred= rgsrSVM.predict(X_test)

In [171]:
y_pred[:10]

array([ 8.87982076, 10.20448626, 12.22871631, 12.10005382, 11.74915701,
        8.87982076,  6.57819393, 10.3934437 , 12.59261938,  7.94916579])

In [172]:
print('MAE:', metrics.mean_absolute_error(Y_test, y_pred))  
print('MSE:', metrics.mean_squared_error(Y_test, y_pred))  
print('RMSE:', np.sqrt(metrics.mean_squared_error(Y_test, y_pred)))

MAE: 5.9411955392380325
MSE: 100.91827244298825
RMSE: 10.045808700298261


**NN- MLPRegressor**

In [165]:
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import StandardScaler

In [166]:
scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train) 
X_test = scaler.transform(X_test)
regr = MLPRegressor(random_state=1, max_iter=500).fit(X_train, Y_train)
y_pred= regr.predict(X_test)

In [167]:
X_train

array([[-0.35077952, -1.23591223, -0.05982693],
       [-0.35077952, -1.00994846, -0.95444605],
       [-0.35077952, -1.12293034,  0.53658581],
       ...,
       [ 0.97746805,  0.23285233,  1.7294113 ],
       [ 1.65140648,  1.13670744,  1.13299856],
       [-0.35077952,  0.34583422,  1.7294113 ]])

In [168]:
y_pred[:10]

array([10.40513178, 12.28250889, 16.13291522, 16.77560351, 14.81514119,
       10.40513178,  6.82536185, 13.53531827, 17.25515726,  9.3695401 ])

In [169]:
from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(Y_test, y_pred))  
print('MSE:', metrics.mean_squared_error(Y_test, y_pred))  
print('RMSE:', np.sqrt(metrics.mean_squared_error(Y_test, y_pred)))

MAE: 6.338889404249081
MSE: 87.86862069660079
RMSE: 9.373826363689526
